File: composeApp/src/androidMain/kotlin/com/example/kmp_basic_app/ui/screens/CharacterDetailScreen.kt
package com.example.kmp_basic_app.ui.screens

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.FavoriteBorder
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import coil3.compose.AsyncImage
import com.example.kmp_basic_app.domain.model.CharacterStatus
import com.example.kmp_basic_app.ui.components.ErrorView
import com.example.kmp_basic_app.ui.components.LoadingIndicator
import com.example.kmp_basic_app.ui.theme.NotionColors
import com.example.kmp_basic_app.viewmodel.CharacterDetailViewModel
import org.koin.compose.viewmodel.koinViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CharacterDetailScreen(
    characterId: String,
    onBackClick: () -> Unit,
    viewModel: CharacterDetailViewModel = koinViewModel()
) {
    val state by viewModel.uiState.collectAsStateWithLifecycle()

    LaunchedEffect(characterId) {
        viewModel.loadDetail(characterId)
    }

    // Hero image entry animation
    val heroAlpha = remember { Animatable(0f) }
    val heroScale = remember { Animatable(1.05f) }
    val contentOffset = remember { Animatable(40f) }

    LaunchedEffect(state.detail) {
        if (state.detail != null) {
            heroAlpha.animateTo(1f, tween(400))
        }
    }
    LaunchedEffect(state.detail) {
        if (state.detail != null) {
            heroScale.animateTo(1f, spring(dampingRatio = Spring.DampingRatioLowBouncy, stiffness = Spring.StiffnessLow))
        }
    }
    LaunchedEffect(state.detail) {
        if (state.detail != null) {
            contentOffset.animateTo(0f, spring(dampingRatio = Spring.DampingRatioNoBouncy, stiffness = Spring.StiffnessMediumLow))
        }
    }

    // Favorite heart bounce
    val heartScale by animateFloatAsState(
        targetValue = if (state.detail?.character?.isFavorite == true) 1.0f else 0.85f,
        animationSpec = spring(
            dampingRatio = Spring.DampingRatioMediumBouncy,
            stiffness = Spring.StiffnessMedium
        ),
        label = "detailHeartScale"
    )

    Scaffold(
        containerColor = MaterialTheme.colorScheme.background,
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        state.detail?.character?.name ?: "Character Detail",
                        style = MaterialTheme.typography.titleLarge
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onBackClick) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                },
                actions = {
                    state.detail?.let { detail ->
                        IconButton(
                            onClick = { viewModel.toggleFavorite() },
                            modifier = Modifier.graphicsLayer {
                                scaleX = heartScale
                                scaleY = heartScale
                            }
                        ) {
                            Icon(
                                imageVector = if (detail.character.isFavorite) Icons.Filled.Favorite else Icons.Filled.FavoriteBorder,
                                contentDescription = "Toggle favorite",
                                tint = if (detail.character.isFavorite) NotionColors.StatusDead else MaterialTheme.colorScheme.onSurface
                            )
                        }
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background,
                    titleContentColor = MaterialTheme.colorScheme.onBackground
                )
            )
        }
    ) { padding ->
        when {
            state.isLoading -> LoadingIndicator(modifier = Modifier.padding(padding))
            state.error != null -> ErrorView(
                message = state.error!!,
                onRetry = { viewModel.loadDetail(characterId) },
                modifier = Modifier.padding(padding)
            )
            state.detail != null -> {
                val detail = state.detail!!
                val character = detail.character
                Column(
                    modifier = Modifier
                        .padding(padding)
                        .fillMaxSize()
                        .verticalScroll(rememberScrollState())
                ) {
                    // Hero image with fade + scale entry
                    AsyncImage(
                        model = character.imageUrl,
                        contentDescription = character.name,
                        contentScale = ContentScale.Crop,
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(300.dp)
                            .graphicsLayer {
                                alpha = heroAlpha.value
                                scaleX = heroScale.value
                                scaleY = heroScale.value
                            }
                    )

                    // Content slides up
                    Column(
                        modifier = Modifier
                            .padding(16.dp)
                            .offset { IntOffset(0, contentOffset.value.toInt()) }
                            .graphicsLayer { alpha = heroAlpha.value }
                    ) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            val statusColor = when (character.status) {
                                CharacterStatus.ALIVE -> NotionColors.StatusAlive
                                CharacterStatus.DEAD -> NotionColors.StatusDead
                                CharacterStatus.UNKNOWN -> NotionColors.StatusUnknown
                            }
                            Box(
                                modifier = Modifier
                                    .clip(RoundedCornerShape(9999.dp))
                                    .background(statusColor.copy(alpha = 0.15f))
                                    .padding(horizontal = 12.dp, vertical = 4.dp)
                            ) {
                                Text(
                                    text = character.status.name.lowercase().replaceFirstChar { it.uppercase() },
                                    style = MaterialTheme.typography.labelSmall,
                                    color = statusColor
                                )
                            }
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(
                                text = "${character.species} - ${character.gender}",
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }

                        Spacer(modifier = Modifier.height(24.dp))

                        Text(
                            "Origin",
                            style = MaterialTheme.typography.labelLarge,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(
                            text = detail.origin.name,
                            style = MaterialTheme.typography.bodyLarge,
                            color = MaterialTheme.colorScheme.onBackground
                        )
                        if (detail.origin.type.isNotBlank() || detail.origin.dimension.isNotBlank()) {
                            Text(
                                text = listOf(detail.origin.type, detail.origin.dimension)
                                    .filter { it.isNotBlank() }
                                    .joinToString(" - "),
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }

                        Spacer(modifier = Modifier.height(16.dp))

                        Text(
                            "Location",
                            style = MaterialTheme.typography.labelLarge,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(
                            text = detail.location.name,
                            style = MaterialTheme.typography.bodyLarge,
                            color = MaterialTheme.colorScheme.onBackground
                        )
                        if (detail.location.type.isNotBlank() || detail.location.dimension.isNotBlank()) {
                            Text(
                                text = listOf(detail.location.type, detail.location.dimension)
                                    .filter { it.isNotBlank() }
                                    .joinToString(" - "),
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }

                        Spacer(modifier = Modifier.height(24.dp))

                        Text(
                            "Episodes (${detail.episodes.size})",
                            style = MaterialTheme.typography.labelLarge,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                        LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            items(detail.episodes) { episode ->
                                Box(
                                    modifier = Modifier
                                        .clip(RoundedCornerShape(9999.dp))
                                        .background(MaterialTheme.colorScheme.primaryContainer)
                                        .padding(horizontal = 12.dp, vertical = 6.dp)
                                ) {
                                    Text(
                                        text = "${episode.episode}: ${episode.name}",
                                        style = MaterialTheme.typography.labelSmall,
                                        color = MaterialTheme.colorScheme.onPrimaryContainer
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/Sources/Screens/CharacterDetailScreen.swift
import SwiftUI
import Shared

struct CharacterDetailScreen: View {
    let characterId: String
    @StateObject private var viewModel: CharacterDetailViewModel
    @State private var heroVisible = false
    @State private var contentVisible = false

    init(characterId: String) {
        self.characterId = characterId
        _viewModel = StateObject(wrappedValue: CharacterDetailViewModel(characterId: characterId))
    }

    var body: some View {
        Group {
            if viewModel.isLoading {
                LoadingView()
            } else if let error = viewModel.error {
                ErrorView(message: error) {
                    Task { await viewModel.loadDetail() }
                }
            } else if let detail = viewModel.detail {
                detailContent(detail)
            }
        }
        .navigationTitle(viewModel.detail?.character.name ?? "Detail")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                if let detail = viewModel.detail {
                    Button {
                        Task { await viewModel.toggleFavorite() }
                    } label: {
                        Image(systemName: detail.character.isFavorite.boolValue ? "heart.fill" : "heart")
                            .foregroundColor(detail.character.isFavorite.boolValue ? .red : NotionColors.textTertiary)
                            .symbolEffect(.bounce, value: detail.character.isFavorite.boolValue)
                    }
                }
            }
        }
        .task {
            await viewModel.loadDetail()
        }
    }

    @ViewBuilder
    private func detailContent(_ detail: CharacterDetail) -> some View {
        ScrollView {
            VStack(spacing: 0) {
                // Hero image with fade + scale entry
                AsyncImage(url: URL(string: detail.character.imageUrl)) { phase in
                    switch phase {
                    case .success(let image):
                        image
                            .resizable()
                            .aspectRatio(contentMode: .fill)
                            .scaleEffect(heroVisible ? 1.0 : 1.05)
                            .opacity(heroVisible ? 1 : 0)
                            .animation(.spring(response: 0.5, dampingFraction: 0.8), value: heroVisible)
                            .onAppear {
                                withAnimation {
                                    heroVisible = true
                                }
                            }
                    case .failure:
                        Color(NotionColors.warmWhite)
                            .overlay(
                                Image(systemName: "person.crop.circle")
                                    .font(.system(size: 60))
                                    .foregroundColor(NotionColors.textTertiary)
                            )
                    case .empty:
                        Color(NotionColors.warmWhite)
                            .overlay(ProgressView().tint(NotionColors.accent))
                    @unknown default:
                        EmptyView()
                    }
                }
                .frame(height: 300)
                .clipped()

                // Content slides up
                VStack(alignment: .leading, spacing: 20) {
                    HStack {
                        Text(detail.character.name)
                            .font(.title.bold())

                        Spacer()

                        statusBadge(detail.character.status)
                    }

                    VStack(alignment: .leading, spacing: 12) {
                        infoRow(label: "Species", value: detail.character.species)
                        infoRow(label: "Gender", value: detail.character.gender)
                    }
                    .padding(16)
                    .notionCard()

                    if let origin = detail.origin {
                        sectionHeader("Origin")
                        VStack(alignment: .leading, spacing: 6) {
                            Text(origin.name)
                                .font(.body)
                            if !origin.type.isEmpty {
                                Text(origin.type)
                                    .font(.subheadline)
                                    .foregroundColor(NotionColors.textSecondary)
                            }
                        }
                        .padding(16)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .notionCard()
                    }

                    if let location = detail.location {
                        sectionHeader("Last Known Location")
                        VStack(alignment: .leading, spacing: 6) {
                            Text(location.name)
                                .font(.body)
                            if !location.type.isEmpty {
                                Text(location.type)
                                    .font(.subheadline)
                                    .foregroundColor(NotionColors.textSecondary)
                            }
                        }
                        .padding(16)
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .notionCard()
                    }

                    let episodes = detail.episodes as? [Episode] ?? []
                    if !episodes.isEmpty {
                        sectionHeader("Episodes (\(episodes.count))")
                        ScrollView(.horizontal, showsIndicators: false) {
                            HStack(spacing: 8) {
                                ForEach(episodes, id: \.id) { episode in
                                    VStack(alignment: .leading, spacing: 4) {
                                        Text(episode.episode)
                                            .font(.caption.bold())
                                            .foregroundColor(NotionColors.badgeText)
                                        Text(episode.name)
                                            .font(.caption)
                                            .lineLimit(1)
                                    }
                                    .padding(.horizontal, 12)
                                    .padding(.vertical, 8)
                                    .background(NotionColors.badgeBg)
                                    .clipShape(Capsule())
                                }
                            }
                        }
                    }
                }
                .padding(20)
                .offset(y: contentVisible ? 0 : 30)
                .opacity(contentVisible ? 1 : 0)
                .animation(.spring(response: 0.5, dampingFraction: 0.85).delay(0.15), value: contentVisible)
                .onAppear {
                    contentVisible = true
                }
            }
        }
    }

    private func statusBadge(_ status: CharacterStatus) -> some View {
        HStack(spacing: 6) {
            Circle()
                .fill(statusColor(for: status))
                .frame(width: 8, height: 8)
            Text(status.name)
                .font(.subheadline.weight(.medium))
                .foregroundColor(NotionColors.badgeText)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 6)
        .background(NotionColors.badgeBg)
        .clipShape(Capsule())
    }

    private func infoRow(label: String, value: String) -> some View {
        HStack {
            Text(label)
                .font(.subheadline)
                .foregroundColor(NotionColors.textSecondary)
                .frame(width: 80, alignment: .leading)
            Text(value)
                .font(.subheadline)
        }
    }

    private func sectionHeader(_ title: String) -> some View {
        Text(title)
            .font(.headline)
            .padding(.top, 4)
    }

    private func statusColor(for status: CharacterStatus) -> Color {
        switch status {
        case .alive: return NotionColors.statusAlive
        case .dead: return NotionColors.statusDead
        default: return .gray
        }
    }
}
